### DISCUSSION TITLE
Add option to exclude extra fields when exporting a model.

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#3508] [Feature Request]
When working with a model that has the Config setting `extra: 'allow'`, the export functionalities `dict` and `json` include all extra fields that were supplied when instanciating the model from a dictionary.

In that situation, exporting becomes a blackbox, since there are unknown fields in the model. If the user has no control over the fields present in the source dictionary, those cannot easily be excluded. There is no possible list that could be passed with the regular `exclude` argument.

A workaround is to explicitly `include` all fields of the model, which is very verbose. By code this can (to my knowledge) only be done by accessing the internal `__fields__` dictionary, which usually should be avoided.

So what I will be requesting is an additional parameter `exclude_extra` to the functions `dict` and `json` that will not consider any attributes that were not defined in the model declaration.

Example usage:
```python
class Foo(BaseModel):
    a: int

    class Config:
        extra = 'allow'

Foo(**{'a': 1, 'b': 2}).dict()                           # == {'a': 1, 'b': 2}
Foo(**{'a': 1, 'b': 2}).dict(exclude_extra=True)         # == {'a': 1}
```

**Comment by ghost** [comment]
I am prepared to contribute an implementation for it, see https://github.com/krohdebayes/pydantic/blob/9b1506277f17799520f472ddfa24663ac97f90e8/pydantic/main.py#L863-L864 for main logic.

**Comment by samuelcolvin** [comment]
I was thinking today we should change how extra works significantly in v2, I'll try to describe it properly tomorrow.

Therefore I wouldn't bother making the change how.

**Reply by ghost** [reply]
Are you willing to share some of the thoughts? Are those behaviour changes on the internal implementation side or functionality wise?

**Comment by dnlopes** [comment]
Do we have any news on this topic? This would be a handy feature.

**Comment by dnlopes** [comment]
In my case, I have a fairly complex hierarchy of objects that all inherit from `BaseModel`, and I sometimes I have the need to create one of those objects from a dict which might contain several other fields.

This is what I did, and it seems to work:

```python
class RootObject(BaseModel):
    pass

_T = TypeVar("_T", bound=RootObject)

class MyObject(RootObject):
    my_field: str

def clean_extra_fields_from(obj_dict: dict, target_obj_class: Type[_T]) -> dict:
    class CleanModel(target_obj_class):
        class Config:
            extra = Extra.ignore
    
    cleaned_obj = CleanModel(**obj_dict)
    return target_obj_class(**cleaned_obj.dict()).dict()
```

Now I can pass to this function a dirty `dict` with extra fields and get back another `dict` without the extra fields, respecting the model definition passed as the second argument.

**Comment by Mograbi** [comment]
is there an update regarding this feature ?

**Comment by lig** [comment]
@Mograbi I've created an issue #6150 from this discussion. This is not part of the upcoming Pydantic v2 release but will be considered for one of the releases to follow. PRs are welcome of cause.
